doc/design/status_port.md

```diff
@@ -0,0 +1,453 @@
+object in the shared memory management segment. The pointer is acquired in the
+same manner as previous publishers and subscribers via RouDi's unix domain
+socket. The only difference is that one `StatusPortData` object is shared between
+`StatusPortReader` and `StatusPortWriter`. `store()` takes a `function_ref` to
+be able to manipulate the type in-place in shared memory.
+`StatusPortReader` is a template as well and has the same constructor. The first
+and single argument of`take()` is a `function_ref`, which is executed as long as
+the `take()` was unsuccessful. Furthermore, the `StatusPortReader` has the methods
+necessary to be attachable to a `popo::Listener` (e.g. `{enable,disable}Event`).
+
+`StatusPortData` stores the latest atomic `Transaction` as well as two pointers
+to `SharedChunk`s in the shared memory payload segment. The lifetime of the two
+chunks is bound to the lifetime of the `StatusPortData` object. To be able to to
+associate the correct readers and writers a `ServiceDescription` is used.
+
+#### Frankenstein object corner cases
+
+How can we ensure in `StatusPortWriter::store()` that there is no `StatusPortReader`
+reading the chunk to which we want to write to? This could be for example a slow
+`StatusPortReader`. Is a reference counter needed? No, this is not the case. Instead
+the `StatusPortReader` checks if his view of on the world (aka the two chunks)
+has changed in-between.
+
+To illustrate this, let's assume that that there is one `StatusPortWriter` and one
+`StatusPortReader` in this scenario. Further assume that the type transferred via
+the `StatusPort` is
+
+```cpp
+T = cxx::string<36>
+```
+
+The following samples are transmitted by the `StatusPortWriter`
+
+1. "Time And Relative Dimension In Space"
+2. "Twin Ion Engines Fighter"
+3. "USS Enterprise"
+
+After the second transaction the `StatusPortReader` see's the latest transaction
+below and starts reading the `chunk[1]` aka `SECOND`.
+
+```text
+Latest transaction in shared memory managment segment
++----------------------------------+
+|       ActiveChunk::SECOND        |
+|       abaCounter: 1              |
++----------------------------------+
+```
+
+Now the the `StatusPortWriter` starts writing concurrently the third sample and
+overtakes the slower `StatusPortReader`.
+Due the latest transaction depicted below, `writePosition` is calculated to
+`chunks[1]` aka `SECOND`. Below you can find the location `^` at which the writer
+is currently writing and the reader is reading. As a result the strings are mixed
+up at this very moment in time, a so called Frankenstein object. While in this
+case, the reader might still read uncorrupted data if he finished before the
+writer, one can easily imagine that this can lead to very subtile, nasty bugs.
+To avoid this problem the `StatusPortReader` compares the initial transaction
+with the latest transaction and  re-calls the callable as long as they are not
+the same.
```

Comment:
This is why only self contained data is allowed. Lets assume instead of a string a pointer to some data is stored in here.
Then it is possible that the reader either dereferences a no longer valid pointer or a pointer which is completely corrupted. Furthermore containers like vector and string are also strictly forbidden. Assume the following example.
```cpp
myStateport.take([](const cxx::vector & v){
  for(auto iter = v.begin(); iter != v.end(); ++iter) {
    // while iterating the writer clears the vector. when we are already at position 1 we have `iter == v.end() + 1` and this means we will iterate until we reach an unaccessible memory location and segfault.
  }
});
```

This means every data structure with any kind of logic may cause a catastrophic failure. Therefore only plain old data, c arrays or structs of them should be allowed. And even here we have some serious problems. Look at this one here
```cpp
template<uint64_t N>
using charArray = char[N];

myStateport.take([](const charArray<100> & helloWorld){
  std::cout << helloWorld << std::endl;
  // lets assume the charArray contains "hello world\0" and is updated from the writer to "bla\0 fuu asldalsdkalsdkalsdasd"
  // this piece of code would segfault since there is no final "\0" anymore and the writer may assume that a reader only prints "bla" since it is terminated with "\0". I know it is a little bit constructed but nevertheless one have to be aware of the extreme restrictions we have here.
});
```

Reply:
In short the usage of any class you find in `cxx` is forbidden as type.

Reply:
You are completely right. `read()` should only be used to very simple data types. For more complex data types I'd suggest to add a 2nd method `copy()`, which does a `memcpy` into a local object and checks if the data is un-corrupted before passing it to the `callable`.

Reply:
Without copying (memcpy and verify) before performing any operation the data type is basically constrained to be `trivially copyable` and all its operations (if the exist) must be `lock-free thread-safe`, which is fairly restrictive.

Still, it should work with structs with POD at least.

Reply:
@elfenpiff Ok, got your point and agree. I added the hint to the design doc, but leave the `copy` code snippet in for now. To easily make the question vividly when implementing it, is that ok for you?